File: fibonacci/fib.c
/**
 * Fibonacci Sequence
 *
 * Calculates n digits of the fibonacci sequence, where n is a command line
 * argument to the program.
 */

#include <stdio.h>
#include <stdlib.h>

long long *calculateFibonacciSequence(int n)
{
  if (n <= 0) {
    return 0;
  }

  int i = 0;
  unsigned long long *numbers = malloc(sizeof(long long) * (n + 1));

  if (!numbers) {
    return 0;
  }

  numbers[0] = 0;
  numbers[1] = 1;

  for (i = 1; i < n; i++) {
    numbers[i + 1] = numbers[i] + numbers[i - 1];
  }

  return numbers;
}

int main(int argc, char *argv[])
{
  int i, n;
  unsigned long long *sequence;

  if (argc <= 1) {
    puts("Please pass the number of iterations to run");
    return 1;
  }

  n = atoi(argv[1]);

  if (n <= 0) {
    puts("Please provide a valid number of iterations to run");
    return 1;
  }

  if (n > 93) {
    puts("This program is only capable of producing 93 fibonacci numbers");
    return 1;
  }

  sequence = calculateFibonacciSequence(n);

  if (sequence) {
    for (i = 0; i <= n; i++) {
      printf("%d: %llu\n", i, sequence[i]);
    }

    free(sequence);
  } else {
    puts("Error processing fibonacci sequence");
    return 1;
  }

  return 0;
}

File: fibonacci/fib.php
#!/usr/bin/env php
<?php

/**
 * Calculate the fibonacci sequence to $n digits, starting at $start. Mostly
 * putting this in a function so that HipHopVM can optimize it efficiently (it
 * won't optimize global code).
 *
 * @param  integer $n
 * @param  integer $start
 * @return array
 */
function calculateFibonacciSequence($n, $start = 0)
{
    // Need to calculate at least one number
    if ($n <= 0) {
      throw new InvalidArgumentException('Expected $n to be greater than zero, got ' . $n);
    }

    // Can't start with a number under zero
    if ($start !== 0 && $start !== 1) {
      throw new InvalidArgumentException('Expected $start to be 0 or 1, got ' . $start);
    }

    // Get the seed values
    if ($start === 0) {
        $numbers = [0, 1];
    } else {
        $numbers = [1, 1];
    }

    // Calculate the sequence
    for ($i = 1; $i < $n; $i++) {
        $numbers[] = $numbers[$i] + $numbers[$i - 1];
    }

    return $numbers;
}

// Check to see how many arguments we got
if ($argc <= 1) {
    echo "Please pass the number of iterations to run.\n";
    exit(1);
}

// Get the number of arguments to run
$n = (int) $argv[1];

// Did they specify a start value
if (isset($argv[2])) {
    $start = (int) $argv[2];
} else {
    $start = 0;
}

try {
    // Calculate the sequence
    $sequence = calculateFibonacciSequence($n, $start);

    // Concatenate the array of numbers with ',' as a separator
    echo implode(',', $sequence) . "\n";
} catch (Exception $e) {
    echo $e->getMessage() . "\n";
    exit(1);
}
